ember-client/shared/src/main/scala/org/http4s/ember/client/EmberConnection.scala

```diff
@@ -16,17 +16,41 @@
 
 package org.http4s.ember.client
 
-import cats._
 import cats.effect.Concurrent
 import cats.effect.Resource
+import cats.effect.kernel.Deferred
 import cats.effect.kernel.Ref
+import cats.effect.std.Hotswap
 import cats.syntax.all._
+import fs2.Chunk
 
 private[ember] final case class EmberConnection[F[_]](
     keySocket: RequestKeySocket[F],
+    chunkSize: Int,
     shutdown: F[Unit],
     nextBytes: Ref[F, Array[Byte]],
-)(implicit F: MonadThrow[F]) {
+    hotRead: Hotswap[F, Deferred[F, Either[Throwable, Option[Chunk[Byte]]]]],
+    nextRead: Ref[F, Deferred[F, Either[Throwable, Option[Chunk[Byte]]]]],
+)(implicit F: Concurrent[F]) {
+  def isValid: F[Boolean] = {
+    val isOpen = keySocket.socket.isOpen
+    val isEof = nextRead.get.flatMap(_.tryGet).map {
+      case Some(result) => result.fold(_ => true, _.isEmpty) // if Left or None this socket is dead
+      case None => false // no read yet, which is good!
+    }
+    (isOpen, isEof).mapN((open, eof) => open && !eof)
```

Comment:
Since `mapN` would be strict on both arguments, is it worth using a `flatMap` to short-circuit sooner? If so, which of the two conditions is more likely to fail first?

Reply:
Eh, they are both (volatile) reads, so not particularly expensive, and on the happy-path there is no short-circuiting. So I would rather not sacrifice readability for the unusual, not-so-happy path.